code/go/0chain.net/smartcontract/zcnsc/models.go

```diff
@@ -0,0 +1,586 @@
+		if err != nil {
+			return err
+		}
+		mp.Amount = state.Balance(*value)
+	}
+
+	id, ok = objMap["receiving_client_id"]
+	if ok {
+		if id == nil {
+			return errors.New("receiving_client_id is missing in the payload")
+		}
+		var value *string
+		err = json.Unmarshal(*id, &value)
+		if err != nil {
+			return err
+		}
+		mp.ReceivingClientID = *value
+	}
+
+	id, ok = objMap["signatures"]
+	if ok {
+		if id == nil {
+			return errors.New("signatures is missing in the payload")
+		}
+		var sigs []*json.RawMessage
+		err = json.Unmarshal(*id, &sigs)
+		if err != nil {
+			return err
+		}
+
+		for _, raw := range sigs {
+			sig := &AuthorizerSignature{}
+			err = json.Unmarshal(*raw, sig)
+			if err != nil {
+				return err
+			}
+			mp.Signatures = append(mp.Signatures, sig)
+		}
+	}
+
+	return err
+}
+
+func (mp *MintPayload) GetStringToSign() string {
+	return encryption.Hash(fmt.Sprintf("%v:%v:%v:%v", mp.EthereumTxnID, mp.Amount, mp.Nonce, mp.ReceivingClientID))
+}
+
+func (mp *MintPayload) verifySignatures(ans *AuthorizerNodes) (err error) {
+	signatureScheme := chain.GetServerChain().GetSignatureScheme()
+	toSign := mp.GetStringToSign()
+	for _, v := range mp.Signatures {
+		if v.ID == "" {
+			return errors.New("authorizer ID is empty in a signature")
+		}
+
+		if ans.NodeMap[v.ID] == nil {
+			return errors.New(fmt.Sprintf("authorizer %s not found in authorizers", v.ID))
+		}
+
+		key := ans.NodeMap[v.ID].PublicKey
+		_ = signatureScheme.SetPublicKey(key)
+
+		if key == "" {
+			return errors.New("authorizer public key is empty")
+		}
+
+		ok, err := signatureScheme.Verify(v.Signature, toSign)
+		if !ok || err != nil {
+			return err
+		}
+	}
+
+	return
+}
+
+type BurnPayloadResponse struct {
+	TxnID           string `json:"0chain_txn_id"`
+	Nonce           int64  `json:"nonce"`
+	Amount          int64  `json:"amount"`
+	EthereumAddress string `json:"ethereum_address"`
+}
+
+func (bp *BurnPayloadResponse) Encode() []byte {
+	buff, _ := json.Marshal(bp)
+	return buff
+}
+
+func (bp *BurnPayloadResponse) Decode(input []byte) error {
+	err := json.Unmarshal(input, bp)
+	return err
+}
+
+type BurnPayload struct {
+	Nonce           int64  `json:"nonce"`
+	EthereumAddress string `json:"ethereum_address"`
+}
+
+func (bp *BurnPayload) Encode() []byte {
+	buff, _ := json.Marshal(bp)
+	return buff
+}
+
+func (bp *BurnPayload) Decode(input []byte) error {
+	err := json.Unmarshal(input, bp)
+	return err
+}
+
+type AuthorizerParameter struct {
+	PublicKey string `json:"public_key"`
+	URL       string `json:"url"`
+}
+
+func (pk *AuthorizerParameter) Encode() (data []byte, err error) {
+	data, err = json.Marshal(pk)
+	return
+}
+
+func (pk *AuthorizerParameter) Decode(input []byte) error {
+	err := json.Unmarshal(input, pk)
+	return err
+}
+
+type AuthorizerNode struct {
+	ID        string                    `json:"id"`
+	PublicKey string                    `json:"public_key"`
+	Staking   *tokenpool.ZcnLockingPool `json:"staking"`
```

Comment:
It is not a good idea to organize data as a tree without proper normalization to be honest. 
Every change to pool for example, 
```
transfer, resp, err = ans.NodeMap[tran.ClientID].Staking.EmptyPool(gn.ID, tran.ClientID, tran)
```
will cause changing of AuthorizerNodes. 
- it is serialization/deserialization of pretty large json
- it is contention on all list instead of small it's part

It'll be better to normalize storage and use Ids of all nested entities, fortunately all of them do have one. Instead of storing entity or list we will store id or list of ids, nested entities will be stored in separate MPT node as independent structure. 

This approach has several drawbacks though: 
 - more read queries to MPT which shouldn't be a problem
 - bigger read set and wider MPT
 
But the biggest benefit is that when we change state, we touch narrow part of it, since addition of authorizers won't happen too frequently, but manipulations on their state like pools will. We had similar discussion with @Sriep.  
 
 Do you think such a normalization makes sense?

Reply:
In that case how we should address the same issue in **miners smart contract** and **interestpool smart contract** 

https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/minersc/models.go#L587

`// MinerNode struct that holds information about the registering miner.
type MinerNode struct {
	*SimpleNode `json:"simple_miner"`
	Pending     map[string]*sci.DelegatePool `json:"pending,omitempty"`
	Active      map[string]*sci.DelegatePool `json:"active,omitempty"`
	Deleting    map[string]*sci.DelegatePool `json:"deleting,omitempty"`
}`

Reply:
I definitely suggest addressing it the same way. @platsko made good attempt to change the way we store lists in magmasc, I do not like local db for lists tbh, but the direction is correct.

Probably, we should extract this refactoring to  another task and make separate PR for this changes in all smart contracts at once. @guruhubb what do you think?

Reply:
I agree we can scale to big number as per https://0chain.slack.com/archives/C020E8HA8TE/p1630528421006600 results that @platsko got.  @peterlimg @Sriep thoughts?

Reply:
A good idea, I was horrified when I first saw [getNodesList](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/minersc/models.go#L1204). 

But where is the balance between extra MPT reads, and unmarshalling data unnecessarily?

Reply:
I think authorizers, miners, sharders, consumer lists will be small and fixed. The providers list is extensive and thats why @platsko chose that way.  Perhaps we can use this for blobbers.

Reply:
@guruhubb @Sriep @dabasov I'm taking into account all suggestions and I will rework this smart contract according to magma sc

Reply:
I will implement lists in this PR, though is seems already huge

Reply:
@Sriep We discussed with @guruhubb that issue with MPT update.
I'm making refactoring regarding how we register authorisers but it won't be like with magmasc, there are some other issues under consideration. As far as MPT update is concerned, the issue with multiple updates and MPT queries doesn't affect much the performance so I decided to make refactoring in a separate branch, meanwhile adding benchmarks to the authorisers. That way will be easier to estimate the effect of refactoring

Reply:
So we can merge the request now

Reply:
@Sriep Waiting for your approval

Reply:
@peterlimg Could you review please?